src/Microsoft.AspNet.Security.OAuth/OAuthBearerAuthenticationHandler.cs

```diff
@@ -24,87 +35,167 @@ protected override AuthenticationTicket AuthenticateCore()
             return AuthenticateCoreAsync().GetAwaiter().GetResult();
         }
 
-        protected override async Task<AuthenticationTicket> AuthenticateCoreAsync()
+		/// <summary>
+		/// Searches the 'Authorization' header for a 'Bearer' token. If the 'Bearer' token is found, it is validated using <see cref="TokenValidationParameters"/> set in the options.
+		/// </summary>
+		/// <returns></returns>
+		protected override async Task<AuthenticationTicket> AuthenticateCoreAsync()
         {
+			ExceptionDispatchInfo authFailedEx = null;
+			OAuthRequestTokenContext requestTokenContext = null;
             try
             {
                 // Find token in default location
-                string requestToken = null;
-                string authorization = Request.Headers.Get("Authorization");
-                if (!string.IsNullOrEmpty(authorization))
+				requestTokenContext = new OAuthRequestTokenContext(Context, null);
+
+				// Give application opportunity to find from a different location, adjust, or reject token
+				var messageReceivedNotification =
+					new MessageReceivedNotification<OAuthRequestTokenContext, OAuthBearerAuthenticationOptions>(Context, Options)
+					{
+						ProtocolMessage = requestTokenContext,
+					};
+
+				// notification can set the token
+				await Options.Notifications.MessageReceived(messageReceivedNotification);
+				if (messageReceivedNotification.HandledResponse)
+				{
+					return messageReceivedNotification.AuthenticationTicket;
+				}
+
+				if (messageReceivedNotification.Skipped)
+				{
+					return null;
+				}
+
+				if (string.IsNullOrEmpty(requestTokenContext.Token))
```

Comment:
AFAICT, the access token is expected to be changeable from the `MessageReceived` notification.